### PR TITLE
feat: add regular tree grammars (0.25 Cohort 4 Task 21)

### PR DESCRIPTION
**Pull request by justinelliottcobb**
# 0.25 Cohort 4, Task 21: regular tree grammars

The authored grammar surface over the Task 18–20 automaton core.
RED-first: `tests/regular_tree_grammar.rs` (14 tests + property).

## API

- `Nonterminal` / `GrammarProduction` / `RegularTreeGrammar` —
  checked construction throughout: rank mismatch at production
  construction; undeclared lhs/child/start, unrenderable names,
  duplicate declarations, and ceilings at grammar construction
  (new `RewriteError::MalformedGrammar`).
- **Fixed text syntax** (`amari-tree-grammar/v1` header,
  `nonterminals:`/`start:` declarations, `S -> f A B` productions;
  per-production rank ⇒ overloaded symbols expressible).
  `render()` canonical; `parse()` total with line-numbered typed
  errors; `parse(render(g)) == g`.
- **Lossless conversions** `to_automaton` /
  `from_automaton[_with_limits]`. Documented normalization: the
  automaton alphabet is derived from production symbols, so unused
  declared alphabet symbols are dropped by a round trip — everything
  else preserved exactly (property-tested byte-identical).
- `accepts` / `witness` / `language_is_empty` route through the
  automaton conversion — no duplicated algorithms.

## Property

80 seeded random automata (alphabet includes a binary symbol):
round trips byte-identical after normalization, render/parse fixed
point, membership parity on 20 random terms per instance.

## Evidence

- rewrite matrix: 195 / 195 / 208 / 208 / 230, 0 failed
- discovery: 1116 passed / 0 failed
- clippy `-D warnings`, both CI invocations, **fresh
  `CARGO_TARGET_DIR`**: 0 errors
- MSRV 1.89 check; rustdoc `-D warnings`; fmt clean
- catalog regenerated post-fmt: 11,033 items, hash `c95f1080`
- catalog shard 20/20; sharding/publish-order/workflow verifiers PASS

### CODE REVIEW

File: amari-rewrite/src/language/grammar.rs [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+                .collect();
+            let arity = u16::try_from(children.len())
+                .map_err(|_| malformed(line_number, "rank exceeds u16".into()))?;
+            let symbol = RankedSymbol::new(Symbol::new(tokens[2]), arity);
+            productions.push(
+                GrammarProduction::new(lhs, symbol, children)
+                    .map_err(|_| malformed(line_number, "rank mismatch".into()))?,
+            );
+        }
+        Self::new(nonterminals, productions, starts, limits).map_err(|error| match error {
```
Comment (justinelliottcobb) [inline amari-rewrite/src/language/grammar.rs]: **[P1] Enforce parse limits before allocating the entire grammar.**

All productions are allocated before `Self::new()` checks the supplied ceiling. Nonterminal/start declarations and per-production token/child vectors likewise grow unchecked. Consequently, very small caller limits do not bound parsing memory or work.

**Confirmed reproducer:** parse a valid header plus `nonterminals: S`, `start: S`, and two million `S -> a` lines, using `TreeAutomatonLimits::new(1, 1, 1)`. Under a 256 MiB address-space cap the process aborts on allocation instead of returning `InvalidLimit`. The input should be rejected as soon as the second production exceeds the one-production limit; buffering the remaining lines is unnecessary.

**Suggested fix:** enforce nonterminal and production counts incrementally before allocation/push, reject duplicate starts as they are read, and bound tokenization/child collection by the rank ceiling rather than collecting an arbitrarily large line first. Preserve typed limit errors. Add a large-input/tiny-limit regression to verify early rejection.

File: amari-rewrite/src/language/grammar.rs [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+            }
+        }
+        for production in &productions {
+            if production.symbol().arity() as usize > limits.max_rank() {
+                return Err(RewriteError::InvalidLimit {
+                    resource: "tree automaton rank",
+                    value: usize::from(production.symbol().arity()),
+                    ceiling: limits.max_rank(),
+                });
+            }
```
Comment (justinelliottcobb) [inline amari-rewrite/src/language/grammar.rs:219]: **[P2] Validate or escape every name needed for a lossless render/parse round trip.**

The production validation checks rank and references but not terminal-symbol renderability. Separately, `Nonterminal::is_renderable()` accepts directive-prefixed names even though `parse()` dispatches them through `strip_prefix("start:")` / `strip_prefix("nonterminals:")` before recognizing productions.

**Two confirmed cases:**

1. Construct a grammar with nonterminal/start S and one nullary production whose symbol name is `"a S"`. Construction succeeds and it accepts the constant with that name. Rendering emits `S -> a S`; reparsing succeeds but creates unary `a(S)` instead. The original nonempty language becomes empty. Empty or newline-containing symbol names pose related round-trip problems.
2. Construct a grammar with nonterminal/start `"start:S"` and a nullary a production. Construction succeeds, but the rendered production `start:S -> a` is treated as a start declaration and reparsing fails.

**Suggested fix:** either add an unambiguous escaping/token syntax or reject unsupported symbol names and reserved directive-prefixed nonterminal names during checked construction. Ensure the same rule is applied to automaton-to-grammar conversion. Add round-trip tests covering whitespace/empty terminal names and directive-prefixed nonterminals.

File: amari-rewrite/src/language/grammar.rs [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+    /// automaton→grammar→automaton round trip (unused symbols are
+    /// dropped); states, transitions, and finals are preserved
+    /// exactly.
+    pub fn to_automaton(&self) -> RewriteResult<TreeAutomaton> {
+        let mut alphabet: Vec<RankedSymbol> = self
+            .productions
+            .iter()
+            .map(|production| production.symbol().clone())
+            .collect();
+        alphabet.sort();
```
Comment (justinelliottcobb) [inline amari-rewrite/src/language/grammar.rs:317]: **[P2] Reconcile documented symbol overloading with the automaton's single-rank invariant.**

The grammar documentation explicitly permits overloaded ranks, and construction/parsing accepts them, but collecting both ranked symbols here and passing them to `TreeAutomaton::new()` produces a rank-conflict error. Every grammar language operation routes through this conversion, so the supposedly valid grammar is unusable.

**Confirmed reproducer:** parse `amari-tree-grammar/v1
nonterminals: S
start: S
S -> f
S -> f S
`. Parsing succeeds, but `accepts(&Term::constant("f"))` returns `MalformedAutomaton { message: "rank conflict for symbol f: 0 vs 1" }` instead of accepting the base production.

**Suggested fix:** either implement ranked-symbol identity consistently through conversion and membership, or explicitly reject per-name rank conflicts in `RegularTreeGrammar::new()`/`parse()` with a typed grammar error and remove the overload-support claim. Add a test exercising language operations on this mixed-rank example, not just successful parsing.

**Review by justinelliottcobb — COMMENTED** [review]
Review of 18809bd: three confirmed findings, reproduced in a temporary snapshot without modifying the working checkout. The existing --no-default-features suite passes (195 tests). Three semantic reproduction tests fail, and a separate memory-capped parser reproduction aborts instead of returning the requested limit error.

**Comment by justinelliottcobb** [comment]
All three findings addressed in `c240795`, each RED-first with the reviewer's counterexamples reproduced in-process (6 new regression tests; 20 total):

**P1 — parse limits before allocation.** `parse()` now enforces ceilings *incrementally*: nonterminal/production counts are checked before insertion into bounded sets, duplicate declarations/starts/productions are typed errors as read, and production tokenization collects children lazily, rejecting the line as soon as it exceeds the rank ceiling. The 2M-line/one-production reproducer now returns `InvalidLimit` from the second *distinct* production (memory bounded); a 100k duplicate-line input errors at line 5 (`MalformedGrammar`, duplicates are now constructor errors too — consistent with nonterminals/starts); a 1M-child line errors at the 17th child without collecting the rest.

**P2 — render/parse fidelity.** The renderable-name rule now covers terminal **symbols** (production construction rejects empty / whitespace-containing / arrow names — the `"a S"` symbol that silently changed the language on reparse is rejected) and rejects **directive-prefixed** nonterminal names (`start:S`, `nonterminals:…`) that would be misparsed as declaration lines. Applied uniformly through `from_automaton`.

**P2 — overload contradiction.** Each symbol name now has ONE rank per grammar (matching the automaton alphabet's single-rank invariant); mixed-rank use is a typed `MalformedGrammar` at construction/parse instead of a surprise rank conflict inside `to_automaton`. The overload claim is removed from the docs; the mixed-rank example now fails at parse with a grammar error, pinned by a test.

Evidence: rewrite matrix 201/201/214/214/236 (five feature configs); discovery 1116/0; fresh-target clippy `-D warnings` ×2 clean; MSRV 1.89; rustdoc `-D warnings`; fmt clean; catalog regenerated post-fmt (hash `e7686063`); catalog shard 20/20.

**Review by justinelliottcobb — COMMENTED** [review]
Re-review of c240795: the name-validation and rank-conflict fixes work, and the original oversized-production reproducer now fails early under the same memory cap. The existing --no-default-features suite passes (201 tests). One remaining parser allocation gap is reproduced below. All verification used a temporary snapshot; the working checkout was not modified.

File: amari-rewrite/src/language/grammar.rs [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+                    }
+                    declared.insert(nonterminal);
+                }
+                continue;
+            }
+            if let Some(rest) = line.strip_prefix("start:") {
+                let mut found = false;
+                for token in rest.split_whitespace() {
+                    let start = Nonterminal::new(token);
+                    if !start_set.insert(start) {
```
Comment (justinelliottcobb) [inline amari-rewrite/src/language/grammar.rs]: **[P1] Bound distinct start declarations before inserting them.**

Duplicate starts are now rejected, but arbitrarily many distinct starts can still be inserted into `start_set` without checking `limits.max_states()`. Validation of undeclared references happens only in the final constructor, so the parser can exhaust memory before returning a typed error even with a one-state limit.

**Confirmed reproducer:** parse the header, `nonterminals: S`, and three million distinct lines `start: q0` through `start: q2999999`, with `TreeAutomatonLimits::new(1, 1, 1)`. Under a 256 MiB address-space cap, input construction completes successfully (46,888,928 bytes), then the parser aborts on allocation instead of rejecting the excess starts.

**Suggested fix:** after checking for a duplicate, check the distinct-start count against `max_states` before constructing/inserting another owned entry. This bound is valid even when starts precede declarations: a valid grammar cannot have more distinct starts than its maximum number of declared nonterminals. Return a typed limit error on the first excess start. Add a large-distinct-starts regression alongside the duplicate-start and production-limit tests; the duplicate check alone does not cover this path.

**Comment by justinelliottcobb** [comment]
Fixed in `bb1a2dd` (see commit for full evidence), RED-first with a 200k-distinct-starts regression under a one-state limit.

`start_set` was duplicate-checked but never count-bounded — the parser buffered every distinct start before construction rejected them. Since starts ⊆ declared nonterminals (enforced by construction), a valid grammar never has more distinct starts than the state ceiling, so the parser now checks the distinct-start count against `max_states` *before* inserting each owned entry and returns `InvalidLimit` on the first excess start — the reviewer's reproducer shape now errors from the second line in well under a second.

Evidence: rewrite matrix 202/202/215/215/237; discovery 1116/0 (one transient doctest *linker* bus error during validation was tmpfs pressure from my own accumulated fresh-target clippy dirs — freed 8G, re-ran clean); fresh-target clippy `-D warnings` ×2 clean; MSRV 1.89; rustdoc `-D warnings`; fmt clean; catalog hash unchanged (`e7686063` — the catalog hashes public API surfaces, and this change is internal to `parse`); catalog shard 20/20.

**Comment by justinelliottcobb** [comment]
Correction: the fix commit is `b12a132` (typo'd as `bb1a2dd` above).
